### PR TITLE
Danny/task 1533722 default task scheduler dev

### PR DESCRIPTION
**Pull request by dannybtsai**
This PR contains changes to start the event queue task via the Task.Run() method so it is on the default task scheduler and will not interfere with caller's task scheduler as some custom task schedulers might be single threaded and execution can be blocked. The second change is replacing the BlockingCollection with ConcurrentQueue to prevent resource leaks (BlockingCollection has multiple wait handles/semaphores) as it is decided not to dispose of the LRU cache to avoid possible exceptions in a multi-thread environment.

### CODE REVIEW

File: src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs [hunk 1]
```diff
@@ -48,15 +48,16 @@ namespace Microsoft.IdentityModel.Tokens
     /// </summary>
     /// <typeparam name="TKey">The key type to be used by the cache.</typeparam>
     /// <typeparam name="TValue">The value type to be used by the cache</typeparam>
-    internal class EventBasedLRUCache<TKey, TValue> : IDisposable
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1001:TypesThatOwnDisposableFieldsShouldBeDisposable")]
```
Comment (TimHannMSFT) [inline src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs]: Please add the Justification in the SuppressMessage.

File: src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs [hunk 2]
```diff
@@ -386,12 +378,12 @@ private void StartEventQueueTaskIfNotRunning()
                 return;
             }
 
-            // if the task is stopped, set _eventQueueTaskState = EventQueueTaskRunning and start a new task
+            // If the task is stopped, set _eventQueueTaskState = EventQueueTaskRunning and start a new task.
+            // Note: we need to call the Task.Run() to start a new task on the default TaskScheduler (TaskScheduler.Default) so it does not interfere
```
Comment (TimHannMSFT) [inline src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs]: nit: ...does not interfere **with the** caller's TaskScheduler...

File: src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs (hunk 1 above)
Comment (mafurman) [reply src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs]: We usually put suppression messages in the GlobalSuppressions.cs file. Is there a reason why it's in the code?

File: src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs [hunk 3]
```diff
@@ -100,7 +101,7 @@ internal class EventBasedLRUCache<TKey, TValue> : IDisposable
 
         internal EventBasedLRUCache(
             int capacity,
-            TaskCreationOptions options = TaskCreationOptions.LongRunning,
+            TaskCreationOptions options = TaskCreationOptions.None, // default to None instead of LongRunning as LongRunning will always start a task on a new thread instead of ThreadPool
```
Comment (mafurman) [inline src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs]: I remember messing around with the value of this property caused multiple hanging unit tests for us. We'll have to make sure that isn't happening again.

File: src/Microsoft.IdentityModel.Tokens/InMemoryCryptoProviderCache.cs [hunk 4]
```diff
@@ -259,11 +259,6 @@ protected virtual void Dispose(bool disposing)
             if (!_disposed)
             {
                 _disposed = true;
```
Comment (mafurman) [inline src/Microsoft.IdentityModel.Tokens/InMemoryCryptoProviderCache.cs:263]: Do we still need a Dispose() method on this object?

File: src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs [hunk 5]
```diff
@@ -408,24 +400,6 @@ private DateTime SetTaskEndTime()
             return DateTime.UtcNow.AddSeconds(EventQueueTaskExecutionTimeInSeconds);
         }
 
-        /// <summary>
-        /// Pause the timer.
-        /// </summary>
```
Comment (mafurman) [inline src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs:413]: I think I'm missing something here... why were the timer related methods removed?

File: test/Microsoft.IdentityModel.Tokens.Tests/CryptoProviderFactoryTests.cs [hunk 6]
```diff
@@ -1076,9 +1076,9 @@ public void ProviderCache_EnsureNoLeakingTasks()
             foreach (var provider in signingProviders)
                 cache.TryRemove(provider);
 
-            WaitTillTasksComplete(cache, waitTimeInSeconds); // wait for the event queue task to complete
-            Assert.True(cache.TaskCount == 0, $"ProviderCacheTest_EnsureNoLeakingTasks 2: unexpected task count: {cache.TaskCount}, expected: 0");
+            AssertNoHangingingTasks(cache, "ProviderCache_EnsureNoHangingTasks");
 
+            // Dispose() should not throw any exception.
```
Comment (henrik-me) [inline test/Microsoft.IdentityModel.Tokens.Tests/CryptoProviderFactoryTests.cs:1079]: >Dispose [](http://example.com/codeflow?start=15&length=7)

remove ?

File: test/Microsoft.IdentityModel.Tokens.Tests/CryptoProviderFactoryTests.cs [hunk 7]
```diff
@@ -1184,23 +1187,22 @@ public void ProviderCache_EnsureNoLeakingTasks_SecurityTokenHandler_SigningCrede
...
-            WaitTillTasksStarted(cache, waitTimeInSeconds); // wait for the event queue task to start
-            Assert.True(cache.TaskCount == 1, $"ProviderCache_EnsureNoLeakingTasks_SecurityTokenHandler_SymmetricEncryptingCredentials: unexpected task count: {cache.TaskCount}, expected: 1");
+            AssertNoHangingingTasks(CryptoProviderFactory.Default.CryptoProviderCache as InMemoryCryptoProviderCache, "ProviderCache_EnsureNoLeakingTasks_SecurityTokenHandler_SigningCredentials");
+        }
 
-            WaitTillTasksComplete(cache, waitTimeInSeconds); // wait for the event queue task to complete
-            Assert.True(cache.TaskCount == 0, $"ProviderCache_EnsureNoLeakingTasks_SecurityTokenHandler_SigningCredentials: unexpected task count: {cache.TaskCount}, expected: 0");
+        private void AssertNoHangingingTasks(InMemoryCryptoProviderCache cache, string callName)
+        {
+            WaitTillTasksComplete(cache, TaskStopWaitTimeInSeconds(cache.TaskExecutionTimeInSeconds)); // wait for the event queue task to complete
```
Comment (henrik-me) [inline test/Microsoft.IdentityModel.Tokens.Tests/CryptoProviderFactoryTests.cs]: >WaitTillTasksComplete [](http://example.com/codeflow?start=12&length=21)

how much time is pend waiting in our utests?

File: test/Microsoft.IdentityModel.Tokens.Tests/EventBasedLRUCacheTests.cs [hunk 8]
```diff
@@ -73,224 +71,212 @@ public void DoNotRemoveExpiredValues()
...
+                // Only even values should expire.
+                if (i % 2 == 0)
+                    cache.SetValue(i, i.ToString(), DateTime.UtcNow + TimeSpan.FromSeconds(5));
+                else
+                    cache.SetValue(i, i.ToString());
+            }
 
-                Thread.Sleep(5000);
-                cache.RemoveExpiredValues();
+            Thread.Sleep(5000);
```
Comment (henrik-me) [inline test/Microsoft.IdentityModel.Tokens.Tests/EventBasedLRUCacheTests.cs:105]: >5000 [](http://example.com/codeflow?start=25&length=4)

needed?

**Comment by henrik-me** [comment]
image
it's a lot of time for tests to run. is that expected?

File: test/Microsoft.IdentityModel.Tokens.Tests/CryptoProviderFactoryTests.cs (hunk 7 above)
Comment (dannybtsai) [reply test/Microsoft.IdentityModel.Tokens.Tests/CryptoProviderFactoryTests.cs]: Currently, the event queue task will end after "Last_Event Time" + 2 min, but in tests the wait is 1.5 times (3 min) to accommodate any overhead of task start/stop.

File: src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs [hunk 9]
```diff
@@ -100,7 +110,7 @@ internal class EventBasedLRUCache<TKey, TValue> : IDisposable
 
         internal EventBasedLRUCache(
             int capacity,
-            TaskCreationOptions options = TaskCreationOptions.LongRunning,
+            TaskCreationOptions options = TaskCreationOptions.None, // default to None instead of LongRunning as LongRunning will always start a task on a new thread instead of ThreadPool
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs]: nit: can you put comments on the line above, thanks.

**Review by brentschmaltz — APPROVED** [review]

**Pull request merged**